### PR TITLE
Ancient and recent mecha bug fixes

### PR DESCRIPTION
**Pull request by necromanceranne**
## About The Pull Request

Mecha guns actually utilize random spread while firing if ``randomspread = TRUE``. Currently every weapon that isn't the shotguns are always pinpoint even if they would have variance.

Makes bumpsmash and melee attacks in a mech use the same cooldown. The actual speed between bumpsmash melees are the same as before (once every 0.3 seconds) and click melee is the same as well (once every second). However, if you do one or the other, it will put you on cooldown for both. The reason for this is that they're literally just calling the same proc but not respecting each others cooldowns. So we've consolidated this into one cooldown with varying cooldown timers. I don't even think this is the most elegant solution, but I'm not going to make any radical changes of the structure of the code. Fuck that.

**Edit** I forgot to mention this but you have to be in combat mode to bumpsmash as a consequence of the above changes. You're fucking welcome.

Separates out mecha_melee_attack proc on the ``/obj/`` level to instead only ``/obj/structure`` and ``/obj/machinery``, which is the only things we should be attacking in the /obj/ list. I don't even want to know what mechs have been able to punch while this wasn't the case. Probably nothing they should have.

## Why It's Good For The Game

Mechs are a fucking diabolical nightmare of procs and some truly ancient code. Over time, things have gotten worse, as we have no one really actively maintaining some of this consistently. One of these bugs is literally day of mech implementation. I shit you not.

## Changelog
:cl:
fix: Mecha weaponry is capable, for the first time ever, of experiencing recoil. This was an intended mechanic, I promise. The code just literally never worked.
fix: Mecha bump melee attacks and click melee attacks are now on the same cooldown, but have varying cooldown timers. You will always bump attack faster than you will click.
fix: You must be in combat mode to punch objects and to bumpsmash into objects.
fix: Stops mecha being able to punch literally any object and damage them.
code: Tidies up some of the autodoc comments for mech weapons.
/:cl:

### CODE REVIEW

**Comment by githubuser4141** [comment]
So this prevents mechs from being able to punch items on the floor? Why can't it "stomp" the items? What about humans, sime/basic mobs? There's a bug that prevents mechs from damaging some basic or simple mobs with punches at the moment.

**Comment by necromanceranne** [comment]
> So this prevents mechs from being able to punch items on the floor? Why can't it "stomp" the items?

It isn't something necessarily expected to be something you can do. There are a few ways to damage items as is, but directly hitting any object isn't functionality shared by human weapons and simple/basic mob attacks, and probably shouldn't be possessed by mechs. It's just asking for a lot of unintentional interactions where we don't expect something to attack a given object, but for contrived reasons, mechs can. If we keep it to the objects we can be sure we want to have attacked, we can be sure whether or not something is working as intended.

>What about humans, sime/basic mobs? There's a bug that prevents mechs from damaging some basic or simple mobs with punches at the moment.

I don't see any issue reports, nor do I see where in the code that would be occurring. You're going to have to elaborate, because as far as I can tell, mechs will punch any living mob, and that includes simple/basic mobs.

**Comment by alien3301** [comment]
Can't punch the lgun and fire axe in mech madness into dust now. 1984

**Comment by necromanceranne** [comment]
I actually completely forgot one of the changes I made on this PR

So for some reason, bumpsmash just...didn't check for combat mode? Like, at all? So you had literally no control over bumpsmashing into things like you could BEFORE the refactor?

**Comment by necromanceranne** [comment]
the bots playing funny buggers.

File: code/modules/vehicles/mecha/mech_melee_attack.dm [hunk 1]
```diff
@@ -28,10 +31,31 @@
...
+	mecha_attacker.do_attack_animation(src)
+	switch(mecha_attacker.damtype)
+		if(BRUTE)
+			playsound(src, 'sound/weapons/punch4.ogg', 50, TRUE)
+		if(BURN)
+			playsound(src, 'sound/items/welder.ogg', 50, TRUE)
+		else
+			return 0
+	mecha_attacker.visible_message(span_danger("[mecha_attacker] hits [src]!"), span_danger("You hit [src]!"), null, COMBAT_MESSAGE_RANGE)
+	TIMER_COOLDOWN_START(mecha_attacker, COOLDOWN_MECHA_MELEE_ATTACK, mecha_attack_cooldown)
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:51]: I think putting the onus on the target procs to start the cooldowns is a bit stinky. This stuff can easily be forgotten or accidentally misplaced

This proc seems to have a return value but a quick glance says that return value is not read anywhere, why not make it so we start the cooldown if this proc returns a value or something?

Comment (necromanceranne) [reply code/modules/vehicles/mecha/mech_melee_attack.dm:51]: Lemme see what I can do.

Comment (necromanceranne) [reply code/modules/vehicles/mecha/mech_melee_attack.dm]: Was easy enough.

**Comment by necromanceranne** [comment]
fucking bot....

**Pull request merged**